### DISCUSSION TITLE
GL.iNet GL-MT3000 running OpenWrt 23.05.4 with Crusader 0.2

### DISCUSSION THREAD

**Discussion by richb-hanover** [Zoxc/crusader#68] [Wi-Fi routers and access points]
## Test conditions

* Firmware: GL.iNet GL-MT3000 router running OpenWrt 23.05.4 firmware
* Crusader Client GUI client: 0.2 running on macOS 12.7.5 Macbook
* Crusader Server: 0.2 pre-built for aarch64 running on router
* Crusader Server was installed by downloading the pre-built aarch64...musl...Linux binary

## Observations

* Ethernet download looks good, latency under 5msec
* Ethernet upload and both labor to reach full speed and latency increases to 35-40msec
* This is likely because `htop` shows the CPU is at 100% during upload/both tests
* 2.4GHz and 5GHz Wi-Fi tend to have lower maximum throughput, but dramatically lower latency

## Ethernet

![plot 2024-09-04 08 42 20](https://github.com/user-attachments/assets/c0a7cc20-491a-4afb-93d2-5ac9ac29ece1)
![plot 2024-09-04 08 41 28](https://github.com/user-attachments/assets/35abf172-5b22-43c8-9b42-428c175de1c2)

## Wi-Fi 2.4GHz 3ft (1m) - 20MHz channel, RSSI -39dB, 802.11n
![plot 2024-09-04 08 53 27](https://github.com/user-attachments/assets/b7947086-a31d-4677-8cd6-8e2d5ed6429f)
![plot 2024-09-04 08 52 37](https://github.com/user-attachments/assets/b40a65c7-75a3-4600-b6f6-5469c92bb19b)
![plot 2024-09-04 08 51 45](https://github.com/user-attachments/assets/183bc681-4094-4656-9828-b7e56e1660e0)

## Wi-Fi 5GHz 3 ft (1m) - 80MHz channel, RSSI -46dB, 802.11ac 

![plot 2024-09-04 08 56 56](https://github.com/user-attachments/assets/371de8b1-c3b4-4f8f-932d-e1bb1fa1a7ea)
![plot 2024-09-04 08 56 08](https://github.com/user-attachments/assets/8cab74d4-fcd3-4711-b833-505f7ae9bb51)
![plot 2024-09-04 08 55 21](https://github.com/user-attachments/assets/53209bd4-91db-4ec4-9314-e56a0df89cc1)

## Wi-Fi 2.4GHz 25ft (8m) - 20MHz channel, RSSI -55dB, 802.11n

![plot 2024-09-04 09 01 05](https://github.com/user-attachments/assets/e1697036-b262-4ac3-b6d1-bb3c92659ca8)
![plot 2024-09-04 09 00 18](https://github.com/user-attachments/assets/c512cdd8-d69e-4713-8306-328975f32bf8)
![plot 2024-09-04 08 59 30](https://github.com/user-attachments/assets/ed3ce3e8-2446-4574-a728-2d9553033bb6)

## Wi-Fi 5GHz 25 ft (8m) - 80MHz channel, RSSI -72dB, 802.11ac

![plot 2024-09-04 09 04 18](https://github.com/user-attachments/assets/70ec3586-55d6-4dff-8470-50b0c971c1c6)
![plot 2024-09-04 09 03 24](https://github.com/user-attachments/assets/8715d6f9-f6bc-4855-84bf-d176c8a62d50)
![plot 2024-09-04 09 02 38](https://github.com/user-attachments/assets/7185d50e-ebf2-4988-a2c9-f1d701d0d42d)

**Comment by richb-hanover** [comment]
**Update: Added more tests using a Crusader server on an RPi4 on Ethernet (instead of running it on the router). The Macbook Pro running the Crusader GUI was on WiFi**

## Wi-Fi 2.4GHz 3ft (1m) to RPi4 on Ethernet - 8 streams
![plot 2024-09-08 15 53 16](https://github.com/user-attachments/assets/2fffd3af-7c54-403c-82dc-d3f78d8da04e)
![plot 2024-09-08 15 52 25](https://github.com/user-attachments/assets/e27d9b1d-7849-47da-b02e-8d4175dc77f0)

## Wi-Fi 2.4GHz 3ft (1m) to RPi4 on Ethernet - 1 stream
![plot 2024-09-08 15 54 12](https://github.com/user-attachments/assets/d37e96b2-bf30-4aa7-8c2d-b2afcc3011c5)

## Wi-Fi 5GHz 3ft (1m) to RPi4 on Ethernet - 8 streams

![plot 2024-09-08 15 57 44](https://github.com/user-attachments/assets/b185ee9b-dc57-4fd2-9604-dab5027aa4bf)
![plot 2024-09-08 15 56 56](https://github.com/user-attachments/assets/06b5dbae-7ac4-463a-828c-4ad0d389f871)

## Wi-Fi 5GHz 3ft (1m) to RPi4 on Ethernet - 1 stream

![plot 2024-09-08 15 58 43](https://github.com/user-attachments/assets/c92c4ade-bb76-4a47-8b37-8b48d9e2a6ee)

**Comment by mcuee** [comment]
Nice results. Basically the dual core Filogic 820 MT7981B CPU is still a bit slow to run all the loads (including Crusader).

On the other hand, GL-MT6000's quad core Filogic 830 MT7986AV CPU is much faster.

**Comment by richb-hanover** [comment]
Yes - a couple observations:

* Both MT3000 (US$86) vs MT6000 ($159 with $25 off coupon) are good deals  at Amazon
* The OpenWrt firmware drops latency considerably
* The OpenWrt firmware also decreases peak throughput significantly
  * 120 mbps up and down for Vendor firmware at 2.4GHz
  * 60 down, 140mbps up or OpenWrt at 2.4GHz
* On 5GHz
  * ~600 mbps Vendor firmware both ways
  * 400 mbps down, 600 mbps up with OpenWrt

**Reply by mcuee** [reply]
Probably because vendor firmware has the acceleration built-in. OpenWRT official firmware may need some time to get that right.

HFO (HW flow offloading) and WED (Wireless Ethernet Dispatch) are not enabled by default OpenWRT official firmware. The following OpenWRT page says that HFO is not compatible with SQM and WEB will bypass AQL. But then SQM/AQL helps reduce latency. Therefore the user needs to see whether raw speed is more important or latency is more important for their use cases.
https://openwrt.org/toh/gl.inet/gl-mt6000

Maybe you can try enable HFO/WED in the official OpenWRT firmware and then see if the performance matches vendor firmware.
